File: client/java/client/src/main/java/dev/angzarr/client/router/SagaHandlerResponse.java
package dev.angzarr.client.router;

import dev.angzarr.CommandBook;
import dev.angzarr.EventBook;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Response from saga handlers.
 *
 * <p>Sagas produce commands (to send to other aggregates) and optionally
 * events/facts (to inject directly into target aggregates).
 */
public class SagaHandlerResponse {

    private final List<CommandBook> commands;
    private final List<EventBook> events;

    private SagaHandlerResponse(List<CommandBook> commands, List<EventBook> events) {
        this.commands = commands != null ? commands : Collections.emptyList();
        this.events = events != null ? events : Collections.emptyList();
    }

    /**
     * Create an empty response (no commands, no events).
     */
    public static SagaHandlerResponse empty() {
        return new SagaHandlerResponse(Collections.emptyList(), Collections.emptyList());
    }

    /**
     * Create a response with commands only.
     */
    public static SagaHandlerResponse withCommands(List<CommandBook> commands) {
        return new SagaHandlerResponse(new ArrayList<>(commands), Collections.emptyList());
    }

    /**
     * Create a response with events only.
     */
    public static SagaHandlerResponse withEvents(List<EventBook> events) {
        return new SagaHandlerResponse(Collections.emptyList(), new ArrayList<>(events));
    }

    /**
     * Create a response with both commands and events.
     */
    public static SagaHandlerResponse withBoth(List<CommandBook> commands, List<EventBook> events) {
        return new SagaHandlerResponse(new ArrayList<>(commands), new ArrayList<>(events));
    }

    /**
     * Get the commands to send to other aggregates.
     */
    public List<CommandBook> getCommands() {
        return commands;
    }

    /**
     * Check if response has commands.
     */
    public boolean hasCommands() {
        return !commands.isEmpty();
    }

    /**
     * Get the events to inject into target aggregates.
     */
    public List<EventBook> getEvents() {
        return events;
    }

    /**
     * Check if response has events.
     */
    public boolean hasEvents() {
        return !events.isEmpty();
    }
}

File: client/csharp/Angzarr.Client/Router/SagaHandlerResponse.cs
namespace Angzarr.Client.Router;

/// <summary>
/// Response from saga handlers.
/// Contains commands to send to other aggregates and/or events/facts
/// to inject directly into target aggregates.
/// </summary>
public class SagaHandlerResponse
{
    /// <summary>
    /// Commands to send to other aggregates.
    /// </summary>
    public List<Angzarr.CommandBook> Commands { get; set; } = new();

    /// <summary>
    /// Events/facts to inject directly into target aggregates.
    /// </summary>
    public List<Angzarr.EventBook> Events { get; set; } = new();

    /// <summary>
    /// Create an empty response.
    /// </summary>
    public static SagaHandlerResponse Empty() => new();

    /// <summary>
    /// Create a response with commands only.
    /// </summary>
    public static SagaHandlerResponse WithCommands(IEnumerable<Angzarr.CommandBook> commands) =>
        new() { Commands = commands.ToList() };

    /// <summary>
    /// Create a response with events only.
    /// </summary>
    public static SagaHandlerResponse WithEvents(IEnumerable<Angzarr.EventBook> events) =>
        new() { Events = events.ToList() };

    /// <summary>
    /// Create a response with both commands and events.
    /// </summary>
    public static SagaHandlerResponse WithBoth(
        IEnumerable<Angzarr.CommandBook> commands,
        IEnumerable<Angzarr.EventBook> events
    ) => new() { Commands = commands.ToList(), Events = events.ToList() };
}
